src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala

```diff
@@ -1,32 +1,63 @@
 package org.constellation.domain.transaction
 
-import cats.effect.Concurrent
+import cats.effect.{Bracket, Concurrent}
 import cats.effect.concurrent.Semaphore
 import cats.implicits._
 import org.constellation.primitives.TransactionCacheData
 import org.constellation.storage.PendingMemPool
 
-class PendingTransactionsMemPool[F[_]: Concurrent](semaphore: Semaphore[F])
-    extends PendingMemPool[F, String, TransactionCacheData](semaphore) {
+class PendingTransactionsMemPool[F[_]: Concurrent](
+  transactionChainService: TransactionChainService[F],
+  semaphore: Semaphore[F]
+) extends PendingMemPool[F, String, TransactionCacheData](semaphore) {
 
   // TODO: Rethink - use queue
   def pull(maxCount: Int): F[Option[List[TransactionCacheData]]] =
-    ref.modify {
-      case txs if txs.isEmpty =>
-        (txs, none[List[TransactionCacheData]])
-      case txs =>
-        val sorted = sortForPull(txs.values.toList)
-        val (left, right) = sorted.splitAt(maxCount)
-        (right.map(tx => tx.hash -> tx).toMap, left.some)
-    }
+    Bracket[F, Throwable].bracket(ref.acquire)(
+      _ =>
+        ref.getUnsafe
+          .flatMap(
+            txs =>
+              (if (txs.isEmpty) {
+                 none[List[TransactionCacheData]].pure[F]
+               } else {
+                 sortForPull(txs.values.toList).flatMap { sorted =>
+                   val (left, right) = sorted.splitAt(maxCount)
+                   ref.unsafeModify(_ => (right.map(tx => tx.hash -> tx).toMap, if (left.isEmpty) none else left.some))
```

Comment:
Alternatively to `if (left.isEmpty) none else left.some)`

```scala
Option.unless(left.isEmpty)(left)
```

https://www.scala-lang.org/api/2.13.x/scala/Option$.html#unless[A](cond:Boolean)(a:=%3EA):Option[A]